opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -248,84 +139,164 @@ message MetricDescriptor {
   //   SumObserver        Sum(aggregation_temporality=cumulative;is_monotonic=true)
   //   UpDownSumObserver  Sum(aggregation_temporality=cumulative;is_monotonic=false)
   //   ValueObserver      Gauge()
-  oneof type {
-    // TODO: Determine if encoding all possible values in a uint64 bitset
-    // improves performance significantly and propose that if that is the case.
+  oneof data {
```

Comment:
Some thoughts here:
1. MeasurementValueType cannot be completely removed because still needed in places like Histogram, unless we double that type as well. There may be other aggregations like that that will not have to expose different types based on the measurement value type (e.g. Summary that will most likely be re-added).
2. I do agree that having a GaugeInt/GaugeDouble is simple enough, for Sum I am a bit worried to duplicate all the fields.

In general case we will have >5 data points per metrics (I remember this from a previous backend that I used), so indeed we may argue that splitting into int/double data points may save some memory for the cost of maintenance of duplicate code.

Another option is to merge https://github.com/open-telemetry/opentelemetry-proto/pull/172 and simply don't care about the in-memory extra 8 bytes per point (no effect on the marshal/unmarshal or network size).

Any ideas?

My preference would be to simplify the model and merge #172 and keep the current MeasurementValueType as part of the Sum/Gauge/Histogram (not in the metric because it may not apply to some aggregation)

Reply:
>MeasurementValueType cannot be completely removed

Correct. What I wanted to say is to remove the `measurement_value_type` field from Gauge/GaugeInt, not remove the MeasurementValueType enum definition.

>I do agree that having a GaugeInt/GaugeDouble is simple enough, for Sum I am a bit worried to duplicate all the fields.

Sum is currently 5 fields and I suggest to remove 2. Only 3 fields will remain, 2 of which will be the same (duplicate) for SumInt64 and SumDouble. Not big deal unless we think more common fields will be added in the future.

>Another option is to merge #172 and simply don't care about the in-memory extra 8 bytes per point (no effect on the marshal/unmarshal or network size).

That works too. Did you do any benchmarking to see if it can hint to a more performant approach? Other than performance I don't have a strong preference for one of the discussed options, we can use either, both have their own benefits and drawbacks.

Reply:
I promise a follow-up immediately when this is merged to resolve this with benchmark numbers. For the moment the main change proposed here is to unify metadata about aggregation with the data so I want to move forward with that.

Reply:
Filed #206 to track this.